crates/kitsune_p2p/kitsune_p2p/src/gossip/simple_bloom/step_4_com_loop_inner.rs

```diff
@@ -0,0 +1,337 @@
+use super::*;
+use kitsune_p2p_types::codec::*;
+
+pub(crate) async fn step_4_com_loop_inner_outgoing(
+    inner: &Share<SimpleBloomModInner>,
+    tuning_params: KitsuneP2pTuningParams,
+    space: Arc<KitsuneSpace>,
+    ep_hnd: Tx2EpHnd<wire::Wire>,
+    peer_cert: Tx2Cert,
+    how: HowToConnect,
+    gossip: GossipWire,
+) -> KitsuneResult<()> {
+    let gossip = gossip.encode_vec().map_err(KitsuneError::other)?;
+    let gossip = wire::Wire::gossip(space, gossip.into());
+
+    let t = tuning_params.implicit_timeout();
+
+    let con = match how {
+        HowToConnect::Con(con) => {
+            if con.is_closed() {
+                let url = pick_url_for_cert(inner, &peer_cert)?;
+                ep_hnd.get_connection(url, t).await?
+            } else {
+                con
+            }
+        }
+        HowToConnect::Url(url) => ep_hnd.get_connection(url, t).await?,
+    };
+    con.notify(&gossip, t).await?;
+
+    Ok(())
+}
+
+pub(crate) async fn step_4_com_loop_inner_incoming(
+    inner: &Share<SimpleBloomModInner>,
+    con: Tx2ConHnd<wire::Wire>,
+    gossip: GossipWire,
+) -> KitsuneResult<()> {
+    use crate::event::*;
+
+    // parse the message
+    let (send_accept, remote_filter) = match gossip {
+        GossipWire::Initiate(Initiate { filter }) => {
+            let bloom_byte_count = filter.len();
+            tracing::debug!(
+                %bloom_byte_count,
+                "incoming 'Initiate'",
+            );
+
+            (true, filter)
+        }
+        GossipWire::Accept(Accept { filter }) => {
+            let bloom_byte_count = filter.len();
+            tracing::debug!(
+                %bloom_byte_count,
+                "incoming 'Accept'",
+            );
+
+            (false, filter)
+        }
+        GossipWire::Chunk(Chunk { finished, chunks }) => {
+            let chunk_count = chunks.len();
+            tracing::info!(
+                %finished,
+                %chunk_count,
+                "incoming 'Chunk'",
+            );
+
+            // parse/integrate the chunks
+            let futs = inner.share_mut(move |i, _| {
+                if let Some(tgt_cert) = i.initiate_tgt.clone() {
+                    if finished && con.peer_cert() == tgt_cert {
+                        i.initiate_tgt = None;
+                    }
+                }
+
+                let mut futs = Vec::new();
+
+                for chunk in chunks {
+                    for agent in i.local_agents.iter() {
+                        match &*chunk {
+                            MetaOpData::Op(key, data) => {
+                                futs.push(i.evt_sender.gossip(
+                                    i.space.clone(),
+                                    agent.clone(),
+                                    agent.clone(), // TODO - from??
```

Comment:
Will this end up on the conductors handle_incoming because that would be bad if the from_agent is set wrong.

Reply:
Why? It's not like the person we got the gossip from was the author. Does it matter? It's hard to determine who it came from, given the local aggregation I'm doing.